### PR TITLE
[SYCL-MLIR]: Make the first operand of SYCLConstructorOp [MemWrite]

### PR DESCRIPTION
**Pull request by etiotto**
The first operand of the SYCLConstructorOp operation is always a memref representing the SYCL object being constructed. Make it use the MemWrite memory effect interface.

### CODE REVIEW

File: mlir-sycl/lib/Dialect/IR/SYCLOps.cpp [hunk 1]
```diff
@@ -64,11 +64,8 @@ bool SYCLCastOp::areCastCompatible(TypeRange Inputs, TypeRange Outputs) {
 }
 
 LogicalResult SYCLConstructorOp::verify() {
-  if (getNumOperands() == 0)
-    return emitOpError("A sycl::constructor must be passed the object to build "
-                       "as the first argument");
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Dialect/IR/SYCLOps.cpp:69]: lit test `invalid.mlir` fails because this is removed.

Comment (victor-eds) [reply mlir-sycl/lib/Dialect/IR/SYCLOps.cpp:69]: Yep, we should remove that test (not applicable anymore)

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 2]
```diff
@@ -1140,8 +1140,9 @@ MLIRScanner::emitSYCLOps(const clang::Expr *Expr,
           !RD->getName().empty()) {
         std::string Name =
             MLIRScanner::getMangledFuncName(*Func, Glob.getCGM());
-        return Builder.create<mlir::sycl::SYCLConstructorOp>(Loc, RD->getName(),
-                                                             Name, Args);
+        SmallVector<Value> RemainderArgs(Args.begin() + 1, Args.end());
+        return Builder.create<mlir::sycl::SYCLConstructorOp>(
+            Loc, Args[0], RemainderArgs, RD->getName(), Name);
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:1145]: Can we use an `ArrayRef` instead of a `SmallVector` to avoid memory allocation?

**Review by victor-eds — COMMENTED** [review]
LGTM after two comments

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 2 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:1145]: ok

File: mlir-sycl/lib/Dialect/IR/SYCLOps.cpp (hunk 1 above)
Comment (etiotto) [reply mlir-sycl/lib/Dialect/IR/SYCLOps.cpp:69]: removed

**Review by whitneywhtsang — APPROVED** [review]

**Review by victor-eds — APPROVED** [review]
LGTM when pipeline passes

**Pull request merged**